toc/rfc/rfc-draft-v2-app-manifests.md

```diff
@@ -0,0 +1,642 @@
+impedes supporting new features because it is heavily app-centric and laden with
+backwards-compatibility-induced complexity.
+
+In addition to issues with the current schema, app manifests are
+inconsistently-declarative, which is a constant source of user confusion. In
+addition, this makes it difficult to implement more advanced developer
+workflows (e.g. [GitOps](https://www.gitops.tech/)), since some resources and
+configuration values can not be deleted when applying manifests. For example,
+it is not possible to unset environment variable or delete service bindings by
+applying an app manifest.
+
+### App Manifests vs Space Manifests
+
+Historically, manifests have been called "app manifests", even though they can
+contain multiple apps and create/modify non-app resources in a space. Because
+of this, the Cloud Controller internally refers to them as [space
+manifests](https://github.com/cloudfoundry/cloud_controller_ng/blob/cc0728d112ee98faf20432d6ff962726b73dedd5/app/controllers/v3/space_manifests_controller.rb).
+For simplicity and historical continuity, this RFC refers to manifests as "app
+manifests", but this is worth re-evaluating for the final implementation.
+
+### Selection of Relevant GitHub Issues:
+
+1. https://github.com/cloudfoundry/cloud_controller_ng/issues/1334
+1. https://github.com/cloudfoundry/cloud_controller_ng/issues/1358
+1. https://github.com/cloudfoundry/cloud_controller_ng/issues/2996
+1. https://github.com/cloudfoundry/capi-release/issues/158
+1. https://github.com/cloudfoundry/cli/issues/1710
+
+## Proposal
+
+### Guiding Principles
+
+The low-level details of the v2 app manifest schema may change during
+development, but the following should guide those decisions, whenever possible.
+
+#### Mirror v3 API Design
+
+Resources in the manifest should be named and structured similarly to the v3
+API. This will reduce conceptual overhead for app developers and API consumers.
+This has the additional benefit of reducing design overhead when adding new
+features to app manifests, since the API design will serve as a strong
+foundation.
+
+#### Top-Level Resources
+
+Space-level resources should have their own top-level configuration blocks,
+rather than being nested under apps. For example, routes, service instances,
+and service bindings are now top-level nodes, instead of being nested under
+apps.
+
+#### Prefer Extensible Data Types
+
+Prefer using maps and sequences instead of scalar values. This makes it easier to
+extend values with additional configuration in the future. For example, favor
+
+```yaml
+routes:
+  - url: foo.example.com
+```
+
+over
+
+```yaml
+routes:
+  - foo.example.com
+```
+
+and especially
+
+```yaml
+route: foo.example.com
+```
+
+#### Keys Use Underscores
+
+Version 1 app manifests use a combination of underscores and hyphens for
+multi-word keys. Schema version 2 keys should only use underscores, to match
+the v3 API convention.
+
+### Behavior
```

Comment:
**question** Any thoughts on whether v2 manifests mimic the behaviour of v1 where you may have to restart/restage the app in order for configuration to take effect (e.g. update env vars)? I can see it being annoying to apply a v2 manifest and then have to take an imperative action (restart) to actually see changes applied.

Reply:
That will probably remain the same. That gives users control over how they want to restart their app ("hard" restart vs restage vs deployment).